### ISSUE TITLE
Enhance @parser to allow additional dependencies

### ISSUE THREAD

**Issue by csams** [RedHatInsights/insights-core#1888]
We should extend `@parser` to allow additional dependencies that just control whether or not the parser fires. Then we can have different parsers for the same spec, and we can have little components like `is_rhel8`, `is_rhel7` they can depend on.

That ensures any exceptions the parsers raise are still meaningful and not just the result of trying to parse something they could never understand because the data is from the wrong RHEL.

We also still have the option of putting a combiner in front of them when it makes sense.

Approaches I considered:

- Have multiple specs for the same file. This just moves the problem up a level. They'd have to depend on something like is_rhelX, too, and it could mean a different spec and registry point for every single version of something we want to support.
- Allow parsers to depend on other components like RedHatRelease which they then use to select the correct parsing function. This has the advantage of providing a single interface to the data for rules to use, but it complicates the parser, and it forces the same interface to the data, even if there are major differences across versions.
- Allow parsers to depend on other components, but those dependencies just control whether the parser fires. Parsers don't actually get to inspect them. This is conceptually similar to the second approach, but it leans on the framework to the keep the parser implementations simpler and decoupled from one another, and it allows us to present a single interface through a combiner without forcing it where it doesn't make sense.

**Comment by xiangce** [comment]
@csams - If I understand correctly, allowing parsers to depend on other components, is just like the current `combiner` a little, right? For example, in the parser, we do nothing except for passing the `context` to the `combiner`, and in the combiner, the real corresponding `parser logic` can be applied to the `context`.

```python
# Parser
@parser(Specs.rsyslog_conf)
class RsyslogConf(ConfigParser):  # For both rhel-7 and rhel-8
  '''
  .. warn::   # <-- Put a WARNING here, likes: The parser does nothing, please use the combiner instead.
  '''
  pass

# Combiner
@combiner(RedHatRelease, RsyslogConf)
class RsyslogConf():  # For the configuration files on both rhel-7 and rhel-8
  def parse_content_rhel_7(content):   # <-- the real parser
    ...
  def parse_content_rhel_8(content):   # <-- the real parser
    ...
  def __init__(rh_rel, rsl_conf):
    if rh_rel.rhel7:
      parse_content_rhel_7(rsl_conf.content)
    elif rh_rel.rhel8:
      parse_content_rhel_8(rsl_conf.content)
    else:
      raise SkipException()
    # <- handle the other context here, e.g `file_path`, `file_name`, and etc.

        
```

I mean, seems the current `combiner` could achieve the goal, is it necessary to enhance/apply the new feature to the `parser`. (which would make it more complicated)

**Comment by csams** [comment]
@xiangce The idea is to have two parsers like this:
```python
@parser(Specs.rsyslog_conf, is_rhel_7)
class RSyslogConfRhel7(Parser):
    ...

@parser(Specs.rsyslog_conf, is_rhel_8)
class RSyslogConfRhel8(Parser):
    ...

@combiner([RSyslogConfRhel7, RSyslogConfRhel8])
class RSyslogConf(object):
    def __init__(rhel7_conf, rhel8_conf):
        ...
```
The correct parser will fire depending on which `is_rhel_*` is successful. The additional dependencies just control whether the parser fires or not. If the options change significantly between versions, the parser for that version can reflect the changes, and rules can choose which version they want to target. We also can still have a combiner to provide a single interface where that makes sense. This lets us keep the convention that parsers target particular file formats, and combiners can help hide differences when the formats change.

**Comment by xiangce** [comment]
@csams Your resolution makes sense to me. And it looks pretty cleaner and better than implementing it based on the current design. I just thought that if the current framework could handle it, we do not need to enhance anything else😄

**Comment by xiangce** [comment]
I think this issue is resolved by #1920, can we close it?
